docs/sources/userguide/plugins.md

```diff
@@ -0,0 +1,157 @@
+**Request**
+
+```
+{HostPath: "/path",
+ ContainerID: "abcdef123"}
+```
+
+**Response**
+
+```
+{HostPath: "/newpath",
+ ContainerID: "abcdef123"}
+```
+
+In the initial version, if the plugin responds with a 404 then the response is ignored.
+
+See reference implementation:
+
+* [Docker volumes extension mechanism](https://github.com/cpuguy83/docker/compare/ddb366ee9a07e3feab766cc712c9683ad0c3c309)
+* [Flocker extension](https://github.com/clusterhq/powerstrip-flocker/compare/docker-volume-extension)
+
+**Demo**
+
+Despite the name, this demo does not use powerstrip!
+
+```
+$ git clone https://github.com/clusterhq/powerstrip-flocker
+$ cd powerstrip-flocker/vagrant-aws
+$ git checkout docker-volume-extension
+$ vagrant up --provider=aws
+$ vagrant ssh node1
+node1$ docker run -v /flocker/test:/data ubuntu sh -c "echo fish > /data/file.txt"
+node1$ exit
+$ vagrant ssh node2
+node2$ docker run -v /flocker/test:/data ubuntu cat /data/file.txt
+fish
+```
+
+In future, we will want to extend this to notify volume extensions when containers start and stop, so that the underlying volumes mechanism can hold and release *leases* on volumes.
+
+### Volumes CLI
+
+We may want to extend some (as-yet non-existent) docker core volumes CLI in due course, so you can do something like the following:
+
+```
+$ docker volume create --driver flocker \
+	--metadata size=50G,tier=ssd,replication=global nicevol
+$ docker run -v /flocker/nicevol:/data dockerfile/postgresql
+```
+
+This would result in similar API requests being made.
+
+Other `docker volume` commands which could make sense are e.g. `docker volume snapshot` and `docker volume rm` which will need corresponding extension API message types to be defined.
+
+## API
+
+Extensions to the Docker remote API (extension type `api`) can be achieved via a [Powerstrip-like protocol](https://github.com/clusterhq/powerstrip#pre-hook-adapter-endpoints-receive-posts-like-this).
+The corresponding limitations will initially apply to the types of requests which can be processed, but this is acceptable for an MVP.
+
+## Network
+
+Network extensions exist to implement the following UX:
+
+```
+$ docker network create greatnet
+$ docker run --net=greatnet --ip=1.2.3.4 dockerfile/postgresql
+```
+
+(and of course: `$ docker run --net=greatnet --ip=1.2.3.4 -v //nicevol:/data dockerfile/postgresql` should also be possible, when both networking and storage plugins are loaded)
+
+**Implementation note:** Whether `docker network` calls the Docker client binary and Docker daemon, or whether the first class networks functionality is implemented by separate binary *should be transparent to the plugin*.
+
+Similar request-response pairs as for volumes can be defined for network `create`, `add-container`, etc.
+
+
+# Semantics
+
+This area needs fleshing out.
+
+* Should chaining be possible?
+* What happens when you load several of the same type of extension?
+* What happens when a hook or pre-hook fails?
+  What about post-hooks?
```

Comment:
Post hooks are important - IMHO they should certainly be included somehow in the workflow.  I think about it as pre-hook=setup and post-hook=react where the react step is "do something now that container is running"